### ISSUE TITLE
model completion not working for arrays - continued 

### ISSUE THREAD

**Issue by quickbeam123** [Z3Prover/z3#4588]
Thanks for the fix of #4577! It removed around 16% of "bool expression involving arrays not fully evaluated"-problems we are registering.

Another problematic, still remaining, is:

```
  context c;
  solver solver(c);
  
  z3::sort s1 = c.int_sort();
  z3::sort s2 = c.bool_sort();
  z3::sort s = c.array_sort(s1,s2);
  z3::expr e1 = c.constant("csK5",s);
  z3::expr e_arr = c.constant("csK4",s);
  s = c.int_sort();
  z3::expr e_idx = c.constant("csK2",s);
  z3::expr e_val = c.bool_val(false);
  z3::expr e2 = store(e_arr,e_idx,e_val);
  z3::expr e = (e1 == e2);
  z3::expr nm = c.bool_const("v7");
  z3::expr naming = (nm == e);
  
  solver.add(naming);
  cout << "[Z3] add (naming): " << naming << endl;
  
  z3::expr lit1 = e;
    
  z3::expr v7 = nm;
  
  s = c.int_sort();
  e1 = c.constant("cc",s);
  s = c.int_sort();
  e2 = c.constant("csK2",s);
  e = (e1 == e2);
  
  z3::expr lit2 = e;
  z3::expr cl = lit1 || lit2;
  
  solver.add(cl);
  cout << "[Z3] add (clause): " << cl << endl;
  
  s1 = c.int_sort();
  s2 = c.bool_sort();
  s = c.array_sort(s1,s2);
  e_arr = c.constant("csK5",s);
  s = c.int_sort();
  e_idx = c.constant("csK6",s);
  e = select(e_arr,e_idx);
  
  cl = e;
  solver.add(cl);
  cout << "[Z3] add (clause): " << cl << endl;
  
  cout << solver.check() << endl;
  model model = solver.get_model();
  expr assignment = model.eval(v7,true /*model_completion*/);
  cout << "assignment: " << assignment << endl;
```

reporting

```
[Z3] add (naming): (= v7 (= csK5 (store csK4 csK2 false)))
[Z3] add (clause): (or (= csK5 (store csK4 csK2 false)) (= cc csK2))
[Z3] add (clause): (select csK5 csK6)
sat
assignment: (= (store ((as const (Array Int Bool)) false) 1 true)
   (store (store ((as const (Array Int Bool)) false) 1 true) 0 false))

```

Again, we would love `assignment` to be either `true` or `false`.

**Comment by NikolajBjorner** [comment]
Could you try to set rewriter.expand_nested_stores to true?

**Comment by NikolajBjorner** [comment]
Use this one:  set_param("rewriter.expand_store_eq", "true");

**Comment by quickbeam123** [comment]
This indeed helped resolving the above issue. Thank you!

Unfortunately, some instances of the general "model completion not working for arrays" in our test set still remain.

One of them gives:

```Value of assignment is: 
(let ((a!1 (store (store (store ((as const (Array Int 15)) |15!val!24|)
                                2
                                |15!val!3|)
                         0
                         |15!val!5|)
                  7
                  |15!val!10|))
      (a!2 (store (store (store ((as const (Array Int 15)) |15!val!30|)
                                0
                                |15!val!17|)
                         2
                         |15!val!18|)
                  1
                  |15!val!13|)))
(let ((a!3 (store (store (store (store a!2 38 |15!val!21|) 39 |15!val!26|)
                         5
                         |15!val!15|)
                  7
                  |15!val!16|)))
(let ((a!4 (store (store (store (store a!3 7719 |15!val!22|) 6 |15!val!20|)
                         4
                         |15!val!19|)
                  3
                  |15!val!14|)))
  (= (store (store (store a!1 40 |15!val!28|) 4 |15!val!1|) 1 |15!val!8|) a!4))))
```

Pasting it here in hope this could be indicative of the underlying problem. If not, I will try preparing a MWE and possibly submit as new issue. 

Thanks!

**Comment by NikolajBjorner** [comment]
I can see what is going on.

**Comment by quickbeam123** [comment]
Thank you! Things are improving! (The previous one is gone.) Still some remain.

Would the following be indicative of what's the next problem?

```
Value of assignment is: (let ((a!1 (store (store (store ((as const (Array Int 17)) |17!val!271|)
                                91
                                |17!val!153|)
                         54
                         |17!val!78|)
                  128
                  |17!val!227|)))
(let ((a!2 (store (store (store (store a!1 137 |17!val!245|) 82 |17!val!135|)
                         72
                         |17!val!115|)
                  132
                  |17!val!235|)))
(let ((a!3 (store (store (store (store a!2 130 |17!val!231|) 112 |17!val!195|)
                         92
                         |17!val!155|)
                  20
                  |17!val!3|)))
  (= (_ as-array k!20) a!3))))

```

**Comment by NikolajBjorner** [comment]
k!20 is presumably defined in the model but not expanded by the evaluator.

**Comment by quickbeam123** [comment]
Should I prepare an actual example run, or can you think of a fix without one?

More generally, it seems that an "evaluator" is an overkill to our needs. 

We only ever ask `model.eval` on previously introduced (named) expressions of bool sort, which -- I assume -- should correspond to sat variables in the propositional abstraction, and in this sense should be always equal to `true` or `false` in the model. Can one ask about the propositional abstraction directly, instead of talking to an evaluator?

**Comment by quickbeam123** [comment]
Here is a new "MWE" (it's getting hard to minimise these):

```
  context c;
  solver solver(c);
  
  vector<sort> sorts;
  vector<expr> exprs;
  
  set_param("rewriter.expand_store_eq", "true");
  
  // pasting vampire generated stuff:
  #include "z3out.txt"
    
  cout << solver.check() << endl;
  model model = solver.get_model();
  expr assignment = model.eval(c.bool_const("v33"),true /*model_completion*/);
  cout << "assignment: " << assignment << endl;

```
for [z3out.txt](https://github.com/Z3Prover/z3/files/5069807/z3out.txt) returning
```
...
assignment: (let ((a!1 (store (store (store ((as const (Array Int 17)) |17!val!2|)
                                22
                                |17!val!22|)
                         23
                         |17!val!24|)
                  45
                  |17!val!66|)))
(let ((a!2 (store (store (store (store a!1 54 |17!val!84|) 33 |17!val!42|)
                         28
                         |17!val!34|)
                  52
                  |17!val!80|)))
  (= (_ as-array k!9) (store (store a!2 55 |17!val!86|) 13 |17!val!6|))))
```

**Comment by quickbeam123** [comment]
This fixed it, thanks!
